### ISSUE TITLE
`InvalidParameterError` when using MCMCGaussianProcess and scikit-learn >= 1.2.X

### ISSUE THREAD

**Issue by jamesdolezal** [automl/SMAC3#921]
#### Description
Optimizers which use the `MCMCGaussianProcess` will encounter the following error with scikit-learn 1.2.X:

```
sklearn.utils._param_validation.InvalidParameterError: The 'n_restarts_optimizer' parameter of GaussianProcessRegressor must be an int in the range [0, inf). Got -1 instead.
```

The issue is in [`MCMCGaussianProcess._get_gp()`](https://github.com/automl/SMAC3/blob/bf32ca81b8ce06510605159950ec804d90d32c82/smac/epm/gaussian_process/mcmc.py#L279), where the the argument `n_restarts_optimizer` is set to -1, with the comment "Do not use scikit-learn's optimization routine."

In scikit-learn version 1.2.1, the `n_restarts_optimizer` requires a value between 0 and infinity (see [their docs](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html)). In versions <1.2 this was not enforced, but parameters are [now being validated in 1.2.X](https://github.com/scikit-learn/scikit-learn/blob/8c9c1f27b7e21201cfffb118934999025fd50cca/sklearn/gaussian_process/_gpr.py#L222), which is causing the error to be raised.

The solution would be to change:

```
n_restarts_optimizer=-1,  # Do not use scikit-learn's optimization routine
``` 

to 

```
n_restarts_optimizer=0,  # Do not use scikit-learn's optimization routine
``` 

in `smac.epm.mcmc:284`.

**Comment by eddiebergman** [comment]
Hi @jamesdolezal,

We are currently in the process of delivering a new SMAC that's still in development and being trialled in a few projects before being released. In this latest version of SMAC, [this parameter is correctly set!](https://github.com/automl/SMAC3/blob/e5f9875341cc2a2cac69373a31d0352d91581adf/smac/model/gaussian_process/mcmc_gaussian_process.py#L299).

My guess is you're using the `1.4` version of SMAC ([the release history is here](https://pypi.org/project/smac/#history)) which is the current default of `pip install smac`.

In the meantime, you could do a [monkey patch](https://stackoverflow.com/questions/5626193/what-is-monkey-patching) along theses lines:
```python
from smac.epm.gaussian_process.mcmc import MCMCGaussianProcess

def _get_gp(self: MCMCGaussianProcess) -> GaussianProcessRegressor:
  return GaussianProcessRegressor(
      kernel=self.kernel,
      normalize_y=False,
      optimizer=None,
      n_restarts_optimizer=0, # Modified here
      alpha=0,  # Governed by the kernel
  )
MCMCGaussianProcess._get_gp = _get_gp
```

I havn't tried this out properly but this is the strategy I would use if you're bound to `1.4`.

If you are only using the facades of SMAC and it's not business critical, i.e. just optimizing something and not building around it, we would appreciate some beta-testing of our latest releases. `pip install --pre smac` should download the beta-builds. The documentation is online [here](https://automl.github.io/SMAC3) (just removed the `v1.4.0` from the normal documentation link) and the [minimal working example](https://automl.github.io/SMAC3/main/4_minimal_example.html) is a whole lot shorter now.

**Comment by jamesdolezal** [comment]
Thanks for the context regarding the upcoming 2.0 release. Our team is maintaining an [open source project](https://github.com/jamesdolezal/slideflow) that uses SMAC for hyperparameter optimization, so we'll try either the monkey patch approach or require scikit-learn<1.2. I'm happy to try out the beta-builds though - I imagine we'll want to add support once it's stable, anyway.

Is there any interest in fully testing SMAC against scikit-learn 1.2 and providing a 1.4.1 patch compatibility update? Happy to help on that front, if there's interest.

**Comment by eddiebergman** [comment]
I imagine we're fairly compatible with scikit-learn 1.2 and I don't imagine performance has changed much. I'll add it to my agenda for tomorrow to make a 1.4.1 release with this fix then.

The new release is mainly an internal overhaul with a cleaner API and a bit more extensibility, but some functionality may have been lost. If you try it out and this effects you, please let us know!

`slideflow` seems like quite a cool looking tool fyi, thanks for linking us to it! It's very hard to know who uses our tools sometimes. @mlindauer you may want to have a look

**Comment by eddiebergman** [comment]
Hi @jamesdolezal,

Apologies for the delays, we are facing some [slight issues](https://github.com/automl/SMAC3/issues/924) with gpytorch integrations in v1.4.

I added some [hacky work-around](https://github.com/automl/SMAC3/pull/922/commits/d63567266473a205067a339206a30149e24896f9).

This doesn't effect your functionality but I think it's a good sign that building for v1.4 is not going to be the robust solution long term. Hopefully I can get out `v1.4.1` later today!

**Comment by dengdifan** [comment]
This error is fixed under `v1.4.1` https://github.com/automl/SMAC3/pull/922. Feel free to reopen this issue if this error still occurs
